### ISSUE TITLE
SSR

### ISSUE THREAD

**Issue by WickyNilliams** [WickyNilliams/cally#8]
* Is SSR possible?
* If so, how?
* In what environments? 
* Would need docs and examples

I tried `@atomico/astro` previously but hit some issues, and didn't want to go down that rabbit hole before release. Worth revisiting https://github.com/atomicojs/astro

**Comment by roeniss** [comment]
TL;DR: Will SSR be implemented soon?

Hi. I'm trying to introduce Cally to my project. The features and looks perfectly fit in. 

Currently, the Cally Web Component is served as a HTML from my web server with script tag, meaning I have no React-like build process. 

```js
return `
...
<script type="module" src="https://unpkg.com/cally"></script>
<calendar-multi>
    <calendar-month></calendar-month>
</calendar-multi>
...
`
```

and the webpage shows sudden *flash* before fully loaded:

![](https://media1.giphy.com/media/v1.Y2lkPTc5MGI3NjExd2gzZXVnaGx1OG9hZzNjcWh5Z2NrZHU1c3BhZzRyYXVka2d2ZjU4bSZlcD12MV9pbnRlcm5hbF9naWZfYnlfaWQmY3Q9Zw/TwBjTsxlgvlvsXdsun/giphy.gif)

I could reduce the time with make cally.js static file and use it, but I couldn't reduce it to zero.

I think it's impossible to make it zero until SSR implemented. If you have any idea to help, share it please :D

**Comment by WickyNilliams** [comment]
Awesome! You can tweak it further with expected dimensions if you want to avoid layout shift as much as possible

**Comment by dgonzalezr** [comment]
Hello here 👋

I'm also trying to get `cally` to work with Next.js SSR. In my case, I have a components library implemented with [Stencil.js](https://stenciljs.com/docs/introduction) and configured the [stencil output target for SSR](https://stenciljs.com/docs/react#enable-server-side-rendering-ssr).

Before setting up SSR for the components library, I was bundling `cally` with the component by adding `import cally` at the top, which works as expected (with no SSR), but once I started to support SSR, this no longer works, which is also expected since cally rely on DOM API. So, I thought that maybe I could delay the import once the component is running on the client side, something like:

```ts
  async connectedCallback() {
    if (!isClient()) return;

    await import('cally');
  }
```

But still wasn't working as expected with Next.js, since the import gets hoisted.

Then I saw this issue and used the same workaround by injecting the library in the Document's head as a `<script>`, 

```ts
  async connectedCallback() {
    if (!isClient() || isCallyLibraryLoaded()) return;

    try {
      await loadCallyLibrary();
    } catch (error) {
      console.error(error);
    }
  }
```

where `loadCallyLibrary` is a function that creates and appends the script to the `document.head`. 

![Image](https://github.com/user-attachments/assets/9d345477-4b53-419d-8236-b1ec32004e36)

And this kinda works 😆 but I will be very interested in some more robust, future-proof solution if exists.
Is there any recommendation or a better approach I could use? Thank you so so much in advance 🙏

**Comment by WickyNilliams** [comment]
Why does the dynamic import() get hoisted? Is this a stencil thing or a next thing? Bizarre behaviour!

**Comment by WickyNilliams** [comment]
I wonder if I can do something with export conditions so at least it doesn't blow up in server environments. Maybe it should just be a stub module there. But I'm not sure how next bundles client Vs server code

Is marking this particular component as a client component viable or are you trying to avoid that?

**Comment by WickyNilliams** [comment]
Thanks for the info. 

I'm familiar with stencil at least, having built a DS with it before (you might come across me in old issues at some point 😁)

I understand they will analyze the code and split the bundle at import() points. But that shouldn't mean hoisting that call to module level. If your import is conditional (as yours was) it's not a safe change.

I guess another option might be to add an entry point which doesn't self register the component. That way you can control when it happens

Is there any chance you could create a reduced test case with the combo of next/stencil/Cally? I understand that might be a bit of work, but it's much more likely I can dedicate time to it if I have a minimal reproduction of the issue!

**Comment by WickyNilliams** [comment]
Also, please could you add here the stack trace/exception that gets thrown when you encounter the issue (i.e. without any potential workarounds)

**Comment by dgonzalezr** [comment]
> Is there any chance you could create a reduced test case with the combo of next/stencil/Cally? I understand that might be a bit of work, but it's much more likely I can dedicate time to it if I have a minimal reproduction of the issue!

Sure thing, I will set up a minimal repo with a stencil web component (that uses Cally) and a next.js minimal app that uses it.
I'll follow up here once I have it done 🙂

**Comment by WickyNilliams** [comment]
Appreciate it!

**Comment by hronro** [comment]
One approach might be useful for SSR is [Declarative Shadow DOM](https://developer.mozilla.org/en-US/docs/Web/API/Web_components/Using_shadow_DOM#declaratively_with_html), and it's support by all major browsers.

**Comment by SalahAdDin** [comment]
Is it possible?

**Comment by WickyNilliams** [comment]
Declarative shadow DOM is a possibility in the future, but right now the tooling around it isn't good enough.

**Comment by ClarkeFL** [comment]
If anyone is trying to get it to work on Svelte 5. This worked for me. I'm sure it can be improved.

`
<script>
  let showCalendar = $state(false);
  let calendarElement = $state(); 
  onMount(async () => {
  	console.log("Please Mount");
  	if (browser) {
  		try {
  			await import("cally");
  			console.log(" Please load cally.js");
  			showCalendar = true;
  		} catch (e) {
  			console.log("error", e);
  		}
  	}
  });
  
  $effect(() => {
  	if (calendarElement) {
  		calendarElement.addEventListener("change", (e) => {
  			console.log(e.target.value);
  		});
  	}
  });
</script>
{#if showCalendar}
	<calendar-range class="cally p-0" bind:this={calendarElement}>
		<ArrowLeft aria-label="Previous" slot="previous" />
		<ArrowRight aria-label="Next" slot="next" />
		<calendar-month></calendar-month>
	</calendar-range>
{/if}`

**Comment by SalahAdDin** [comment]
Any progress on this?

**Comment by WickyNilliams** [comment]
Unfortunately, no. There still isn't a good general solution for rendering web components server side that I can truthfully recommend. Most solutions are very narrow or fragile. Standards will eventually solve this, I'm sure, but it is a slow process 

In any case, I think this is an example of a component that is fine to be client side only
